### ISSUE TITLE
Authentication for OpenStack nodes via JNLP.

### ISSUE THREAD

**Issue by remzi-e** [jenkinsci/openstack-cloud-plugin#104]
We're using LDAP for authentication on Jenkins server.
For a dumb slave connected via jnlp we can use the secret token provided by Jenkins' node page.
How can we authenticate OpenStack slaves via JNLP without using a plain LDAP username and Password? Is there anyway to get a similar secret from a node created via openstack plugin?

This is the metadata i get on an openstack node when accessing the URL `http://169.254.169.254/openstack/latest/meta_data.json`:
`{
  "availability_zone": "us-west-or1a",
  "hostname": "jenkins-win-utility-slave-1677.novalocal",
  "launch_index": 0,
  "meta": {
    "jenkins-instance": "https://myjenkinsserver.com:12001/",
    "jenkins-template-name": "jenkins-win-utility-slave"
  },
  "name": "jenkins-win-utility-slave-1677",
  "uuid": "3104b256-2554-4215-a956-87a65579cc51"
}`

Is there a way for the Openstack plugin to provide or inject the secret token in the metadata?

**Comment by olivergondza** [comment]
There is a way to pass metadata to both openstack metadata or user data script so jnlp connection can be initiated. Note that it will be disclosed to every person that can query openstack for metadata and users that can access the node itself.

I lean towards putting this into user data script so nodes do not need to query openstack (which might require authentication as well).

**Comment by remzix** [comment]
thanks. looking forward to trying out the solution.

**Comment by rosca** [comment]
@olivergondza For me the user script is not a viable solution, since I want to be able to choose the environment slave.jar runs in. Cloudbase-init runs the user script under cloudbase default user, and the sessionID for the java process is 0 (no GUI attached), since no user has been logged on before script gets the chance to run. This environment is usually not appropriate for running application UI tests.

My solution is to run slave.jar via a scheduled task, running at user login (auto logon) with elevated rights. That makes the java process to be elevated under that user with a valid session ID with GUI. I'm using meta data to pair the machine with the right node.

I would prefer to have an option for exposing the secrets token in the meta data as well.

By the way, thanks for doing this great plugin!

**Comment by olivergondza** [comment]
The plugin already put user data into machine metadata (for informative purposes) so putting the token in dedicated metadata key as well is not any less secure. Looking into the code we should be able to construct the token at the time machine is scheduled (`Computer` does not exist yet). So, does `jenkins-node-secret` sound good enough?

**Comment by rosca** [comment]
@olivergondza Yes, it sounds good. Thanks!

**Comment by pjdarton** [comment]
@rosca At work, we're looking to do exactly what you describe - JNLP Jenkins slave, running on the desktop (hence has access to the desktop session), running "As Administrator" via a scheduled task.  It's nice to know that we're not alone :-)

FYI I've recently been working on shoring up the JNLP support on the plugin and I've (as of yesterday) had JNLP Windows slaves connecting successfully to a Jenkins 2.32 server, including the -secret key, using a VM that was just "Windows that runs cloud-init with Java pre-installed" (i.e. no Jenkins functionality baked into the VM).  Support for this should be in the next version of the plugin.
Personally, I've not yet scripted a User Data script that adds a scheduled task and makes Windows auto login (I've only ever launched the slave "in-line" from OpenStack - all our "using a scheduled task" VMs are static), but I'm pretty sure that what I've done with the plugin will now permit that to be done.

**Comment by pjdarton** [comment]
@rosca I believe that the current code supports what you need, as we're now (successfully) doing what you said you wanted.
We have a scheduled task that starts the slave, giving us control over the environment.
We have a user-data script that creates a properties file in a "known location" that records the various JNLP property values (the URL of the Jenkins server, the value of the -secret argument etc), and then the scheduled task reads that properties file and runs the slave.

The online-help just gives a simple example of running the slave directly from the user-data script *but* you can do things differently if you want.  e.g. our user-data script runs "w32tm /resync", then turns off the windows time service, then does a dhcp renew, then stores the JNLP data into a properties file in C:\J\slave_vars.cmd.  Meanwhile Windows is configured to log in to a static account automatically, and a scheduled task is set to "run on logon" which runs the slave.jar.
We've found this to be highly reliable.
(FYI before we fiddled with the time service and dhcp resync, it was *not* reliable - the Microsoft Windows DHCP Client is incompatible with the Microsoft Windows Time Service and does not cope with the system time being changed, causing a brief network disconnect when the DHCP lease expired, causing the slave connection to be broken)

**Comment by olivergondza** [comment]
@pjdarton, our users turn out to be doing exactly same routine to get Windows working. I am wondering if this is something we should automate/document.

@remzi-e, @rosca, what is there left to do so this will work for you smoothly?

**Comment by pjdarton** [comment]
@olivergondza "Exactly"?  Even the w32tm /resync and DHCP renewal?  I had _hoped_ that such level of extreme-hackery would only be required on my own setup.  If others have found much the same then *yes* we should document this so that others do not need to go through the same painful learning process.

That said, there are limits to the amount we can automate this - we can't provide a publicly-consumable Image of a Windows VM.  We can provide documentation that includes scripts that can be copy-and-pasted, but that's then getting too large to be suited to the online help compiled into the plugin itself - that level of documentation would be better kept on a wiki page.
Personally, I think that the _small_ examples we have in the existing online help are quite large enough and they demonstrate the capabilities of the plugin fairly well.

FYI the technical challenge involved in booting up a Windows VM and running a Jenkins JNLP slave process running is not purely an OpenStack issue.  We also use the vSphere-cloud plugin and we use a similar process for running slaves there as well (but we don't need the w32tm /resync or the DHCP renewal on vSphere).

Perhaps we could create a wiki page detailing these challenges (and solutions that can be copy/pasted) on the Jenkins wiki site and then link to it from both the OpenStack plugin's page and the vSphere-cloud plugin's page (and from any other cloud plugin that starts Windows VMs)?

**Comment by olivergondza** [comment]
Sort of the same, to be exact. @winklerm, can you please evaluate to what extend this is what we do and what part of that are reasonable reusable?

I can envision that we would provide some powershell snippets to get the variable passing, logon tasks etc. easier to reuse... Not sure it is worth it, though.

**Comment by rosca** [comment]
@olivergondza - yes, this seems to have solved the issue. Thank you.

**Comment by winklerm** [comment]
Hello, we indeed do almost the same thing, except for w32tm, and with a difference that we are using Sysprep to instruct the machine to do ``ipconfig /release`` and ``ipconfig /renew`` during specialize phase of Sysprep.

In more detail, our procedure works as follows:
- on a source instance:
 -- we run Sysprep with configuration that sets up auto-logon for specific user
- we take a snapshot of the source instance,
- when a new instance provisioned from this snapshot boots, it does ``ipconfig /release`` and ``ipconfig /renew`` as instructed by Sysprep,
- in cloudbase-init user data, we execute a script that basically stores all required information for JNLP connection, such as Jenkins master node URL, in a defined location like 'C:\jenkins\',
- when the user automatically logs on, a scheduled task is run:
 1. checks if the information stored by user data script is already in 'C:\jenkins' (we ran into trouble when scheduled task ran before cloudbase-init finished so we needed a synchronization point),
 2. downloads the slave agent jar from jenkins master,
 3. runs the slave jar with the arguments stored by user data script.

Probably not needed now, but we also release the IP address just before running Sysprep on the source instance (there were issues without that on Windows 2016 Server) - it was advised to us that it is always better to clean the network stack properly before snapshotting - but it might not be needed now because we added the ipconfig renewal to Sysprep configuration.

As for reusability, it mainly depends on if you are using Sysprep and what configuration you want to pass to it. Perhaps we could prepare a generic Sysprep's Unattend.xml template that would be customizable by the end-user?
